### PR TITLE
ICU-23305 Fix signed integer overflow in NumeratorSubstitution::doParse

### PR DESCRIPTION
**Pull request by FrankYFTang**
#### Checklist
- [X] Required: Issue filed: ICU-23305
- [X] Required: The PR title must be prefixed with a JIRA Issue number. Example: "ICU-NNNNN Fix xyz"
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. Example: "ICU-NNNNN Fix xyz"
- [X] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable
- [X] Approver: Feel free to merge on my behalf

Fixes signed integer overflow in NumeratorSubstitution::doParse and uninitialized members in SymbolMatcher and CombinedCurrencyMatcher.

### CODE REVIEW

**Comment by richgillam** [comment]
This looks fine to me, except maybe for using `double` instead of `int64_t`.  I could have sworn the `int64_t` thing was a fairly recent change to solve some other problem, but I might be misremembering.  You really want to get @grhoten involved here.

**Comment by markusicu** [comment]
> I could have sworn the `int64_t` thing was a fairly recent change to solve some other problem

FYI: This particular code block has been unchanged since Doug ported it from Java in 2005: [link](https://github.com/unicode-org/icu/blame/5fbc5f95852403fca612dafca0e005b015055371/icu4c/source/i18n/nfsubs.cpp#L1344)

**Comment by richgillam** [comment]
> > I could have sworn the `int64_t` thing was a fairly recent change to solve some other problem
> 
> FYI: This particular code block has been unchanged since Doug ported it from Java in 2005: [link](https://github.com/unicode-org/icu/blame/5fbc5f95852403fca612dafca0e005b015055371/icu4c/source/i18n/nfsubs.cpp#L1344)

Okay, then I've got it confused with something else.  Ignore me.

File: icu4c/source/i18n/nfsubs.cpp [hunk 1]
```diff
@@ -1341,17 +1342,17 @@ NumeratorSubstitution::doParse(const UnicodeString& text,
...
-            d *= 10;
+            d *= 10.0;
         }
         // now add the zeros
-        while (zeroCount > 0) {
-            d *= 10;
+        while (zeroCount > 0 && !std::isinf(d)) {
+            d *= 10.0;
             --zeroCount;
         }
```
Comment (markusicu) [inline icu4c/source/i18n/nfsubs.cpp]: Wouldn't it be simpler to keep d as an int64_t and check in the loop condition that it's not `> INT64_MAX / 10`?

Switching to double, which usually has a smaller value range than int64_t, could also limit the set of numbers we are able to parse, right?

File: icu4c/source/i18n/nfsubs.cpp [hunk 2]
```diff
@@ -1341,17 +1342,17 @@ NumeratorSubstitution::doParse(const UnicodeString& text,
...
         // d is now our true denominator
-        result.setDouble(static_cast<double>(n) / static_cast<double>(d));
+        result.setDouble(static_cast<double>(n) / d);
```
Comment (markusicu) [inline icu4c/source/i18n/nfsubs.cpp]: What should we return if d overflows? It looks like we return a bogus value, especially if we use double and it has become infinity. Should we detect the overflow and return, say, NaN?

The old code already didn't deal with things like static_cast<double>(n) overflowing; what happens there if n is out of double range?

File: icu4c/source/test/intltest/itrbnf.cpp [hunk 3]
```diff
@@ -2864,6 +2865,22 @@ IntlTestRBNF::TestTurkishSpellout() {
...
..., 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0001, 0x0000, 0xbc03, 0x3b00, 0x3c00, 0x3000, 0x003b, 0x0078, 0x0030, 0x0078, 0x003a, 0x003e, 0x0025, 0x403e, 0x003e, 0x3e00, 0x3eff, 0x251a, 0x3e00, 0x3e40, 0x3e00, 0x703e, 0x756c, 0x006e, 0x7367, 0x6953, 0x003b, 0x0025, 0x403e, 0x003a, 0x893e, 0x3b00, 0x1000, 0x0000, 0x0000, 0x0000, 0x1000, 0x3c3e, 0x003b, 0x003e, 0x003b, 0x0002, 0x0034, 0x003c, 0x003b, 0x003c, 0x003c, 0x003c, 0x3a40, 0x3e00, 0x0089, 0x003b, 0x0010, 0x0000, 0x0000, 0x0000, 0x3e10, 0x3b3c, 0x3e00, 0x3b00, 0x0200, 0x3400, 0x3c00, 0x3b00, 0x3c00, 0x3c00, 0x3c00, 0x0000, 0x0000, 0x0300, 0x0000, 0x003c, 0x003c, 0x003c, 0x0000, 0x0000, 0x0003, 0x003e, 0x003b, 0x0000, 0x0000, 0x003b, 0x001f, 0x403e, 0x007a, 0x003e, 0x003b, 0x0000, 0x1000, 0x3c3e, 0x003b, 0x003e, 0x003b, 0x0002, 0x0034, 0x003c, 0x003b, 0x003c, 0x003c, 0x003c, 0x0059, 0x0000, 0x0000, 0x0000, 0x1616, 0x3431, 0x0000
```
Comment (markusicu) [inline icu4c/source/test/intltest/itrbnf.cpp]: line breaks please so that the line length is max 100

File: icu4c/source/test/intltest/itrbnf.cpp [hunk 4]
```diff
@@ -2864,6 +2865,22 @@ IntlTestRBNF::TestTurkishSpellout() {
 	}
 }
 
+void
+IntlTestRBNF::TestNumeratorSubstitutionOverflow() {
```
Comment (markusicu) [inline icu4c/source/test/intltest/itrbnf.cpp:2869]: This test function deserves a comment about the test case coming from a fuzzer, and that the goal is to make sure that the code does not crash with UBSan, rather than looking for a particular result value.

File: icu4c/source/i18n/nfsubs.cpp (hunk 1 above)
Comment (grhoten) [reply icu4c/source/i18n/nfsubs.cpp]: I agree with Markus. The bit precision of a double is smaller than int64_t. 

Also I thought that ICU4C should be using uprv_isInfinite instead of std::isinf.

How does ICU4J do this? They should be kept in sync.

File: icu4c/source/i18n/numparse_symbols.h [hunk 5]
```diff
@@ -38,7 +38,7 @@ class U_I18N_API SymbolMatcher : public NumberParseMatcher, public UMemory {
 
   protected:
     UnicodeString fString;
-    const UnicodeSet* fUniSet; // a reference from numparse_unisets.h; never owned
+    const UnicodeSet* fUniSet = nullptr; // a reference from numparse_unisets.h; never owned
```
Comment (grhoten) [inline icu4c/source/i18n/numparse_symbols.h:41]: I like this C++ modernization.

File: icu4c/source/test/intltest/itrbnf.cpp (hunk 3 above)
Comment (grhoten) [reply icu4c/source/test/intltest/itrbnf.cpp]: It would also help if the string was readable. \u escaped characters for null is fine.

**Review by grhoten — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/nfsubs.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/nfsubs.cpp]: Done. Reverted to using `int64_t` for the denominator `d` and added overflow checks in the loops. If it would overflow, we now safely abort and return `false` from `doParse`.

File: icu4c/source/i18n/nfsubs.cpp (hunk 2 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/nfsubs.cpp]: We now return `false` (failure) from `doParse` if overflow is detected, which safely aborts the parsing instead of returning a bogus value.

File: icu4c/source/test/intltest/itrbnf.cpp (hunk 3 above)
Comment (FrankYFTang) [reply icu4c/source/test/intltest/itrbnf.cpp]: Done. Reformatted the test case to have max line length < 100.

File: icu4c/source/test/intltest/itrbnf.cpp (hunk 4 above)
Comment (FrankYFTang) [reply icu4c/source/test/intltest/itrbnf.cpp:2869]: Done. Added a comment explaining that the test case comes from a fuzzer and the goal is to verify that it does not crash or trigger UBSan.

File: icu4c/source/i18n/nfsubs.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/nfsubs.cpp]: Agree. I have reverted to `int64_t` and added overflow checks. This also avoids the need for `std::isinf` / `uprv_isInfinite`.

File: icu4c/source/test/intltest/itrbnf.cpp (hunk 3 above)
Comment (FrankYFTang) [reply icu4c/source/test/intltest/itrbnf.cpp]: Done. Converted the hex array to a `UnicodeString` constructed from concatenated `u"..."` string literals with `\uXXXX` escapes, making it much more readable.

**Comment by FrankYFTang** [comment]
PTAL

**Review by richgillam — DISMISSED** [review]
LGTM

File: icu4c/source/i18n/numparse_symbols.h (hunk 5 above)
Comment (markusicu) [reply icu4c/source/i18n/numparse_symbols.h:41]: These kinds of changes are good, but it's a bit odd to see them in an unrelated pull request.
Can make it harder to cherry-pick a bug fix.

In the future, consider making a separate PR for this kind of cleanup.
Also, adding these initializers probably makes some constructor member initializers redundant. Would be nice to remove those, making such related changes together.

**Review by markusicu — DISMISSED** [review]
Thanks Frank!

@grhoten I believe that Frank has addressed your feedback. Ok if we dismiss your rejection?

File: icu4c/source/i18n/nfsubs.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/nfsubs.cpp]: I have now updated ICU4J (`NFSubstitution.java`) to keep it in sync with ICU4C. Added signed integer overflow checks in `NumeratorSubstitution.doParse` and added `TestNumeratorSubstitutionOverflow` in `RbnfTest.java` (with comments appropriately adjusted for Java exceptions).

File: icu4c/source/test/intltest/itrbnf.cpp [hunk 6]
```diff
@@ -2864,6 +2865,29 @@ IntlTestRBNF::TestTurkishSpellout() {
...
+    // The goal is to make sure that the code does not crash or trigger UBSan
+    // (Undefined Behavior Sanitizer) with invalid/extreme inputs,
+    // rather than looking for a particular result value.
+    UnicodeString rulesStr(
+        u"> \u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
+        u"\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
+        u"\u0001\u0000밃㬀㰀　;x0x:>%䀾>㸀㻿┚㸀㹀㸀瀾畬n獧楓;"
+        u"%䀾:褾㬀က\u0000က㰾;>;\u00024<;<<<㩀㸀\u0089;"
+        u"\u0010\u0000㸐㬼㸀㬀Ȁ㐀㰀㬀㰀㰀㰀\u0000\u0300\u0000<<<\u0000\u0003>;\u0000;"
+        u"\u001f䀾z>;\u0000က㰾;>;\u00024<;<<<Y\u0000ᘖ㐱\u0000",
```
Comment (grhoten) [inline icu4c/source/test/intltest/itrbnf.cpp]: It's so much cleaner. Thank you!

File: icu4c/source/i18n/nfsubs.cpp [hunk 7]
```diff
@@ -1343,10 +1343,16 @@ NumeratorSubstitution::doParse(const UnicodeString& text,
...
+            if (d > INT64_MAX / 10) {
+                return false;
+            }
             d *= 10;
         }
         // now add the zeros
         while (zeroCount > 0) {
+            if (d > INT64_MAX / 10) {
+                return false;
+            }
```
Comment (grhoten) [inline icu4c/source/i18n/nfsubs.cpp:1355]: Why was 10 chosen? Some of the rules use different divisors. Other possibilities include base 20, and base 2.

File: icu4j/main/core/src/main/java/com/ibm/icu/text/NFSubstitution.java [hunk 8]
```diff
@@ -1722,10 +1722,18 @@ public Number doParse(
...
+                    return 0;
+                }
                 d *= 10;
             }
             // now add the zeros
             while (zeroCount > 0) {
+                if (d > Long.MAX_VALUE / 10) {
+                    parsePosition.setIndex(0);
+                    return 0;
+                }
```
Comment (grhoten) [inline icu4j/main/core/src/main/java/com/ibm/icu/text/NFSubstitution.java:1736]: The same comment as the C++ side. Why is this / 10?

**Review by grhoten — COMMENTED** [review]
Can you simplify the test case? The fuzzer test seems to have some irrelevant fluff.

I'm guessing that it's trying to deal with the value 1,555,555,555,555,555,555. That's definitely beyond IEEE754 double precision. It's getting into an area of RBNF that probably needs more range testing between INT64_MAX and the maximum IEE754 precision. I believe that other unit tests for formatting such large numbers, but parsing such values aren't 100% accurate. If that's the case, then this testing is insufficient. You'd essentially be putting a wooden plank over a pothole. It may fix a crash, but it's not actually validating that the results are correct.

I fixed the Number Format Tester not too long ago to support this range of numbers. RBNF is formatting 1,555,555,555,555,555,555 correctly, but it's not parsing it correctly. I'm not sure if that is a tool problem or an RBNF problem. I suspect that it's an RBNF problem.

**Comment by grhoten** [comment]
Actually, I think this is a fraction rule. So it's not a large number. That is also not tested well. I think languages like Russian write such rules. This really needs validation for the expected value and not just if it crashes.

File: icu4c/source/i18n/nfsubs.cpp (hunk 7 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/nfsubs.cpp:1355]: it is testing would the operation next line "d *= 10" overflow first.  

See https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classRuleBasedNumberFormat.html#:~:text=is%20a%20decimal%20number%20expressed%20using%20ASCII%20digits.

bv: | bv specifies the rule's base value. bv is a decimal number expressed using ASCII digits.

File: icu4j/main/core/src/main/java/com/ibm/icu/text/NFSubstitution.java (hunk 8 above)
Comment (FrankYFTang) [reply icu4j/main/core/src/main/java/com/ibm/icu/text/NFSubstitution.java:1736]: because it is testing the following line d*=10 won't be larger than Long.MAX_VALUE

**Comment by FrankYFTang** [comment]
>

**Comment by FrankYFTang** [comment]
> Actually, I think this is a fraction rule. So it's not a large number. That is also not tested well. I think languages like Russian write such rules. This really needs validation for the expected value and not just if it crashes.

The bug I intend to fix here is a fuzzer found security bug that crash. If the code has other problem it should test and report separately. We intend to stop the hacker to attack the user first here.

**Comment by FrankYFTang** [comment]
PTAL

**Comment by FrankYFTang** [comment]
> Can you simplify the test case? The fuzzer test seems to have some irrelevant fluff.

Simplified by AI as below

Yes! I have re-examined the test string in [PR 4057](https://github.com/unicode-org/icu/pull/4057) and simplified it even further down to the absolute mathematical minimum number of \u0000 (NUL) characters required to reproduce the original bug.

1. Why 19 is the Mathematical Minimum
In NumeratorSubstitution::doParse (withZeros == true), the code counts leading zero matches (zeroCount) and multiplies d *= 10 in a loop:

cpp


while (zeroCount > 0) {
    d *= 10;
    --zeroCount;
}
In 64-bit signed integer arithmetic (int64_t in C++, long in Java), INT64_MAX is 9,223,372,036,854,775,807.
10^18 (1,000,000,000,000,000,000) does not overflow INT64_MAX.
To trigger signed integer overflow (which caused the UBSan error in C++ and infinite loop in Java before our fix), d *= 10 must execute at least 19 times to compute 10^19 (10,000,000,000,000,000,000 > INT64_MAX).
Therefore, any test case reproducing the original bug must match at least 19 zeros.
2. How We Simplified the Test String
Eliminated Unnecessary Tail Zeros: In the tail of the rule string (after the initial zeros), we systematically tested removing every single NUL character. Originally there were 19 NULs scattered across various rules in the tail. We proved that 9 of those 10 remaining NULs were completely unnecessary and removed them, leaving only 1 single \u0000 in the tail (inside the >;\u0000; rule).
Reduced Leading Zeros: With exactly 1 NUL being matched in the tail, we only need 18 leading \u0000s at the start of the string (18 + 1 = 19 total zeros matched).
Comparison of Simplification
Original Fuzzer String: 156 characters, containing 64 total \u0000 characters (45 leading zeros + 19 tail zeros).
New Simplified String: 109 characters, containing exactly 19 total \u0000 characters (18 leading zeros + 1 tail zero) — a 70% reduction in \u0000 characters and a 30% reduction in string length.
3. Updated Code & Verification
We updated both C++ (
itrbnf.cpp
) and Java (
RbnfTest.java
) to use this super-simplified string:

java


String rulesStr = "> \u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
    + "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
    + "\u0001밃㬀㰀　;x0x:>%䀾>㸀㻿┚㸀㹀㸀瀾畬n獧楓;"
    + "%䀾:褾㬀ကက㰾;>;\u00024<;<<<㩀㸀\u0089;"
    + "\u0010㸐㬼㸀㬀Ȁ㐀㰀㬀㰀㰀㰀\u0300<<<\u0003>;\u0000;"
    + "\u001f䀾z>;က㰾;>;\u00024<;<<<Yᘖ㐱";
Verified Bug Reproduction: We tested this exact 109-character string against an unrepaired ICU4C build and verified it reproduces the exact original runtime error: runtime error: signed integer overflow: 1000000000000000000 * 10 cannot be represented in type 'long int'

**Review by grhoten — APPROVED** [review]
Fine for now. A real test for this code that validates the actual value should be considered in the future.
